### PR TITLE
added reverse domain name notation to uri

### PR DESCRIPTION
**Pull request by enthus1ast**
added reverse domain name notation to uri

### CODE REVIEW

File: lib/pure/uri.nim [hunk 1]
```diff
@@ -321,6 +321,34 @@ proc `$`*(u: Uri): string =
...
+        if pos < hostname.len:
+            buf.insert ".", 0
+        result.insert(buf,0)
+        buf.setLen 0
+    else:
+        buf.add ch
+    if pos == hostname.len: break
+    pos.inc
+
+proc reverseNotation*(uri: Uri): string =
```
Comment (dom96) [inline lib/pure/uri.nim]: I would prefer it if you just defined one procedure: `reverse(uri: Uri): string`.

Comment (enthus1ast) [reply lib/pure/uri.nim]: but then i need to parseUri the host first.
Maybe this code fits more inside net? 
The code this stems from does not use uri at all for eg.
This is the third time i needed this cause its so much better to sort etc... 
Where would you place it?

**Comment by data-man** [comment]
@enthus1ast 

> ```nim
>         buf.insert ".", 0
>         result.insert(buf,0)
>         buf.setLen 0
> ```

Why do you mix different styles of code writing?

File: lib/pure/uri.nim [hunk 2]
```diff
@@ -321,6 +321,31 @@ proc `$`*(u: Uri): string =
     result.add("#")
     result.add(u.anchor)
 
+proc reverse*(uri: Uri): string =
+  ## returns the `reverse domain name notation 
+  ## <https://en.wikipedia.org/wiki/Reverse_domain_name_notation>`_. 
+  ## of the given hostname.
+  ## A high performance version of this:
```
Comment (dom96) [inline lib/pure/uri.nim]: Add an empty line with ``#`` above and below this line.

```
##
## A high perform...
##
## .. code-block::
```

File: lib/pure/uri.nim [hunk 3]
```diff
@@ -321,6 +321,31 @@ proc `$`*(u: Uri): string =
...
+  ## .. code-block::
+  ##   return hostname.split(".").reversed().join(".")
+  result = ""
+  var
+    pos: int = 0
+    buf: string = ""
+    ch: char
+  while true:
```
Comment (dom96) [inline lib/pure/uri.nim]: Use ``for ch in uri.hostname`` instead of this while loop + ``pos`` + ``ch`` variables.

**Comment by dom96** [comment]
I'm also curious what you're using this for. Are you sure its use is far enough reaching to warrant being included in the stdlib?

**Comment by enthus1ast** [comment]
Generally dns is hierarchically system, but this is "hidden" because of the "forward" notation.
Explicit, lexical sorting! Consider this:

forward & sorted:
```
serv.anothercompany.de
serv.company.com
api.anothercompany.de
api.company.com
```

reversed & sorted:
```
com.company.api
com.company.serv
de.anothercompany.api
de.anothercompany.serv
```


Benchmark:
```

proc reverseNotation*(hostname: string): string =
  ## returns the reverse domain notation of the given hostname
  ##  https://en.wikipedia.org/wiki/Reverse_domain_name_notation
  ## A high performance version of this:
  ##   return hostname.split(".").reversed().join(".")
  result = ""
  var
    pos: int = 0
    buf: string = ""
    ch: char
  while true:
    ch = hostname[pos]
    if ch == '.' or pos == hostname.len:
        if pos < hostname.len:
            buf.insert ".", 0
        result.insert(buf,0)
        buf.setLen 0
    else:
        buf.add ch
    if pos == hostname.len: break
    pos.inc


when isMainModule:
  ## Naive module for performance-testing
  import times

  template timeIt*(name: string, p: untyped): stmt =
    ## Performs timing of the block call, and makes output into stdout.
    let timeStart = cpuTime()
    for idx in 0..10_000:
      p
    echo name, ": ", (cpuTime() - timeStart) * 1000000

  import strutils
  import algorithm
  proc reverseDomain(domain: string): string =
    return domain.split(".").reversed().join(".")


  timeIt "fast":
    discard reverseNotation("foo.baa.foo.baa.foo.baa")

  timeIt "slow":
    discard reverseDomain("foo.baa.foo.baa.foo.baa")
  # echoAssert

when isMainModule:
    echo "server.example.loc".reverseNotation
    echo "server.example.loc.".reverseNotation
    assert  "foo".reverseNotation == "foo"
    assert  "foo.baa".reverseNotation == "baa.foo"
    assert  "foo.baa.baz".reverseNotation == "baz.baa.foo"
    assert  "".reverseNotation == ""
    assert  ".".reverseNotation == "."
    assert  "..".reverseNotation == ".."
    assert "foo.baa".reverseNotation.reverseNotation == "foo.baa"
    assert "server.example.loc".reverseNotation() == "server.example.loc".reverseDomain()

```
fast: 36444.00000000001
slow: 69797.0

**Pull request closed without merging**